jax/interpreters/ad.py

```diff
@@ -191,6 +191,15 @@ def is_linear(var):
                                      map(read_primal, eqn.invars), params)
         map(write_primal, eqn.outvars, ans)
 
+  # Find the last use of each cotangent so that they can be removed
+  # as soon as possible.
+  drop_cts: Dict[int, Any] = {}
+  seen_vars: Set[Any] = set(jaxpr.invars)
+  for eqn in linear_eqns:
+    read_set = set(eqn.outvars)  # NOTE: eqn is not transposed yet!
+    drop_cts[id(eqn)] = read_set - seen_vars
```

Comment:
Another thing that I've assumed is that it is never the case that a single equation object can appear multiple times in the equation list of a jaxpr. But that shouldn't be an issue, because jaxprs are guaranteed to be in SSA form, right?

Reply:
That's my understanding!

Reply:
Could we just use the index of the equation rather than hashing on its id? That is, just make `drop_cts` a list of length equal to the number of equations.

Reply:
Good point! I'll change it as you've suggested